### PR TITLE
[SPARK-13041][Mesos]add driver sandbox uri to the dispatcher ui

### PR DESCRIPTION
**Pull request by skonto**
## What changes were proposed in this pull request?

Adds a sandbox link per driver in the dispatcher ui
Implementation would have been straightforward if it was not for this issue:
https://issues.apache.org/jira/browse/MESOS-4992
As a workaround we have to get runtime info from the master and the slave where the driver was launched on as a task.
There are no scalability issues because we use master/state-summary instead of master/state.json
and ask one slave for the directory of the driver's task.

The sandbox uri has the following format:
`http://172.17.0.1:5050/#/slaves/1d01c68e-96e7-4196-a340-4da8e12892c0-S1/browse?path=%2Ftmp%2Fmesos%2Fslaves%2F1d01c68e-96e7-4196-a340-4da8e12892c0-S1%2Fframeworks%2F1d01c68e-96e7-4196-a340-4da8e12892c0-0009%2Fexecutors%2Fdriver-20160322160503-0001%2Fruns%2Fd90b463e-ad52-433e-bcdc-6c3ffbafdc68`
## How was this patch tested?

Tested with the mesos test suite here: https://github.com/typesafehub/mesos-spark-integration-tests. 
Attached image shows the ui modification.
sandbox

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala [hunk 1]
```diff
@@ -76,6 +104,7 @@ private[mesos] class MesosClusterPage(parent: MesosClusterUI) extends WebUIPage(
       <td>{state.startDate}</td>
       <td>{state.slaveId.getValue}</td>
       <td>{stateString(state.mesosTaskStatus)}</td>
+      {sandboxCol}
```
Comment (dragos) [inline core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala]: Not a big deal, but it looks like all the other cells have `<td>` elements in here, and the call just generates the contents of that cell. I'd prefer to keep the same pattern everywhere.

File: core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala [hunk 2]
```diff
@@ -115,4 +144,58 @@ private[mesos] class MesosClusterPage(parent: MesosClusterUI) extends WebUIPage(
...
+      buffer.array.toList
+    }
+    result.map{byte => byte & 0xFF}.mkString(".")
+  }
+
+  private def getListFromJson(value: JValue): List[Map[String, Any]] = {
+    value.values.asInstanceOf[List[Map[String, Any]]]
+  }
+
+  private def getTasktDirectory(masterUri: String, driverFwId: String, slaveId: String):
```
Comment (dragos) [inline core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala]: Probably a typo, this should be `getTaskDirectory`

**Comment by dragos** [comment]
LGTM apart for minor comments.

File: core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala (hunk 2 above)
Comment (skonto) [inline core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala]: yes i will fix it.

File: core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala (hunk 1 above)
Comment (skonto) [inline core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala]: ok i will fix it.

File: core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala [hunk 3]
```diff
@@ -68,6 +75,25 @@ private[mesos] class MesosClusterPage(parent: MesosClusterUI) extends WebUIPage(
...
+    val sandboxCol = if (masterInfo.isDefined && schedulerFwId.isDefined) {
+
+      val masterUri = masterInfo.map{info => s"http://${getIp4(info.getIp)}:${info.getPort}"}.get
+      val directory = getTaskDirectory(masterUri, id, state.slaveId.getValue)
+
+      if(directory.isDefined) {
+        val sandBoxUri = s"$masterUri" +
+          s"/#/slaves/${state.slaveId.getValue}" +
+          s"/browse?path=${directory.get}"
+          <a href={sandBoxUri}>Sandbox</a>
```
Comment (mgummelt) [inline core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala]: This is going to break DCOS and any other system where the dispatcher sits behind a proxy.  The master is not directly accessible.  You could maybe fix this by adding a proxy prefix property similar to APPLICATION_PROXY_BASE.

File: core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala [hunk 4]
```diff
@@ -68,6 +75,25 @@ private[mesos] class MesosClusterPage(parent: MesosClusterUI) extends WebUIPage(
 
   private def driverRow(state: MesosClusterSubmissionState): Seq[Node] = {
     val id = state.driverDescription.submissionId
+    val masterInfo = parent.scheduler.getSchedulerMasterInfo()
+    val schedulerFwId = parent.scheduler.getSchedulerFrameworkId()
+    val sandboxCol = if (masterInfo.isDefined && schedulerFwId.isDefined) {
+
+      val masterUri = masterInfo.map{info => s"http://${getIp4(info.getIp)}:${info.getPort}"}.get
+      val directory = getTaskDirectory(masterUri, id, state.slaveId.getValue)
```
Comment (mgummelt) [inline core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala]: You're making O(N) HTTP requests.  I think that may be an unacceptable performance cost.

Comment (mgummelt) [inline core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala]: Mesosphere is prioritizing a Mesos fix to make this URL constructable.  I propose we wait for that, then maybe do some version checking here to enable this feature.

**Comment by skonto** [comment]
Hive tests fail...

File: core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala (hunk 3 above)
Comment (skonto) [inline core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala]: Yes i know, but had now way to test it... dont really know the correct format there for the url... please provide the specs in our document.

File: core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala (hunk 4 above)
Comment (skonto) [inline core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala]: Sounds more ugly than it is im afraid, given the options we have. I have already opened a ticket anyway.

**Comment by skonto** [comment]
Jenkins, retest this please.

**Comment by skonto** [comment]
This will be fixed in mesos 0.29.0 release.

**Comment by skonto** [comment]
WIP

**Comment by skonto** [comment]
Not in 0.29 but soon enough. https://issues.apache.org/jira/browse/MESOS-4992

File: core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala (hunk 3 above)
Comment (tnachen) [inline core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala]: I think we should add a property like @mgummelt suggested to override masterUri if available.

File: core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala [hunk 5]
```diff
@@ -68,6 +75,25 @@ private[mesos] class MesosClusterPage(parent: MesosClusterUI) extends WebUIPage(
 
   private def driverRow(state: MesosClusterSubmissionState): Seq[Node] = {
     val id = state.driverDescription.submissionId
+    val masterInfo = parent.scheduler.getSchedulerMasterInfo()
+    val schedulerFwId = parent.scheduler.getSchedulerFrameworkId()
+    val sandboxCol = if (masterInfo.isDefined && schedulerFwId.isDefined) {
+
```
Comment (tnachen) [inline core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala]: Kill space

File: core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala [hunk 6]
```diff
@@ -68,6 +75,25 @@ private[mesos] class MesosClusterPage(parent: MesosClusterUI) extends WebUIPage(
...
+      val masterUri = masterInfo.map{info => s"http://${getIp4(info.getIp)}:${info.getPort}"}.get
+      val directory = getTaskDirectory(masterUri, id, state.slaveId.getValue)
+
+      if(directory.isDefined) {
```
Comment (tnachen) [inline core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala]: Space between if (

File: core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala [hunk 7]
```diff
@@ -68,6 +75,25 @@ private[mesos] class MesosClusterPage(parent: MesosClusterUI) extends WebUIPage(
 
   private def driverRow(state: MesosClusterSubmissionState): Seq[Node] = {
     val id = state.driverDescription.submissionId
+    val masterInfo = parent.scheduler.getSchedulerMasterInfo()
+    val schedulerFwId = parent.scheduler.getSchedulerFrameworkId()
+    val sandboxCol = if (masterInfo.isDefined && schedulerFwId.isDefined) {
+
+      val masterUri = masterInfo.map{info => s"http://${getIp4(info.getIp)}:${info.getPort}"}.get
+      val directory = getTaskDirectory(masterUri, id, state.slaveId.getValue)
+
```
Comment (tnachen) [inline core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala]: Remove extra space

File: core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala [hunk 8]
```diff
@@ -115,4 +142,58 @@ private[mesos] class MesosClusterPage(parent: MesosClusterUI) extends WebUIPage(
...
+    result.map{byte => byte & 0xFF}.mkString(".")
+  }
+
+  private def getListFromJson(value: JValue): List[Map[String, Any]] = {
+    value.values.asInstanceOf[List[Map[String, Any]]]
+  }
+
+  private def getTaskDirectory(masterUri: String, driverFwId: String, slaveId: String):
+      Option[String] = {
+
```
Comment (tnachen) [inline core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala]: Remove extra white spaces here.

File: core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala (hunk 5 above)
Comment (skonto) [inline core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala]: sure

File: core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala (hunk 6 above)
Comment (skonto) [inline core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala]: yes

File: core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala (hunk 7 above)
Comment (skonto) [inline core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala]: ok

File: core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala (hunk 8 above)
Comment (skonto) [inline core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala]: ok

**Comment by skonto** [comment]
@tnachen i will fix those. I am also waiting for the fix.

File: core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala (hunk 3 above)
Comment (skonto) [inline core/src/main/scala/org/apache/spark/deploy/mesos/ui/MesosClusterPage.scala]: Ok i agree i will look into this one.

**Comment by HyukjinKwon** [comment]
Hi @skonto, are you able to proceed this PR further? if not, it might be better closed for now.

**Comment by skonto** [comment]
@HyukjinKwon probably. Let me ask Michael. @mgummelt what do you think? what options do we have?

**Comment by HyukjinKwon** [comment]
gentle ping.

**Comment by HyukjinKwon** [comment]
gentle ping

**Comment by HyukjinKwon** [comment]
@skonto, Probably, how about addressing the comments above and resolve conflicts for now? I think it is better than leaving this open without any further actions. I have been suggesting to close PRs that are inactive to review comments.

I will ping the guys if it is on a mergeable state. Otherwise, I would rather like to close this for now.

**Comment by skonto** [comment]
@HyukjinKwon I will have a look and let you know, please don't close it for now. There was finally progress at the mesos side:
https://reviews.apache.org/r/58872/

**Comment by HyukjinKwon** [comment]
@skonto Sure, I would not propose to close. Thank you for your input.

**Comment by HyukjinKwon** [comment]
@skonto, do you maybe know (or a wild guess) when we would be able to proceed this? Probably, closing this for now and reopening might be an option.

**Comment by haosdent** [comment]
@HyukjinKwon @skonto Sorry for the delay, MESOS-4992 have been fixed.

**Comment by skonto** [comment]
Great thnx @haosdent :)I will proceed with updating the PR. Thnx a lot!

**Comment by skonto** [comment]
@HyukjinKwon I will close this one and start fresh.

**Pull request closed without merging**